File: 0322-coin-change/0322-coin-change.cpp
class Solution {
public:
    int f(vector<vector<int>>& dp , vector<int>& coins , int tar , int i ){
        int n = coins.size();
        if(tar == 0) return 0;
        if(i >= n) return 1e9;
        if(dp[i][tar] != -1) return dp[i][tar];
        int take = 1e9;
        if( coins[i] <= tar){
            take = 1 + f(dp , coins , tar - coins[i] , i);
        }
        int notTake = f(dp , coins , tar , i  + 1);
        return dp[i][tar] = min(take , notTake);
    }
    int coinChange(vector<int>& coins, int amount) {
        int n = coins.size();
        vector<vector<int>> dp(n , vector<int>(amount + 1, -1));
        int ans =  f(dp , coins , amount , 0 );
        return ans == 1e9 ? -1 : ans;
    }
};

File: 0322-coin-change/0322-coin-change.java
class Solution {
    int f(int[][] dp ,int[] coins , int amount , int i){
        int n = coins.length;
        if(i >= n) return (int)1e9;
        if(amount == 0) return 0;
        if(dp[i][amount] != -1) return dp[i][amount];
        int take = (int)1e9;
        if(amount >= coins[i]){
            take = 1 + f(dp , coins , amount - coins[i] , i);
        }
        int notTake = f(dp , coins , amount , i + 1);
        return dp[i][amount] = Math.min(take,  notTake);
    }
    public int coinChange(int[] coins, int amount) {
        int n = coins.length;
        int[][] dp = new int[n][amount + 1];
        for(int [] row : dp){
            Arrays.fill(row , -1);
        }
        int ans = f(dp ,coins , amount , 0);
        return ans == (int)1e9 ? -1 : ans;
    }
}